lib/src/main/java/io/ably/lib/transport/Hosts.java

```diff
@@ -10,18 +10,19 @@
 
 /**
  * Object to encapsulate primary host name and shuffled fallback host names.
+ *
+ * Methods on this class are safe to be called from any thread.
  */
 public class Hosts {
-    private String primaryHost;
-    private String prefHost;
-    private long prefHostExpiry;
-    boolean primaryHostIsDefault;
+    private final String primaryHost;
+    private final boolean primaryHostIsDefault;
     private final String defaultHost;
     private final String[] fallbackHosts;
     private final boolean fallbackHostsIsDefault;
     private final boolean fallbackHostsUseDefault;
     private final long fallbackRetryTimeout;
 
+    private final Preferred preferred = new Preferred();
```

Comment:
I'm a bit surprised you didn't go for the option of `Preferred` being immutable, with `preferred` non-final.

Reply:
Yeah, this was movement of state and the logic around mutating that state, as much for my own benefit to break the moving parts down a bit further... but, you're right, I could still have explored an immutable `Preferred`, as you say, with methods that return a new instance or the same instance as appropriate.

For speed of delivery right now I'm going to leave it as is but I appreciate the observation. 😁